src/multitask_personalization/envs/pybullet/pybullet_skills.py

```diff
@@ -0,0 +1,375 @@
+)
+from multitask_personalization.envs.pybullet.pybullet_structs import (
+    GripperAction,
+    PyBulletAction,
+    PyBulletState,
+)
+from multitask_personalization.rom.models import (
+    ROMModel,
+)
+
+
+def get_object_id_from_name(object_name: str, sim: PyBulletEnv) -> int:
+    """Get the PyBullet ID in the given sim env from a name."""
+    if object_name.startswith("book"):
+        idx = int(object_name[len("book") :])
+        return sim.book_ids[idx]
+    return {
+        "cup": sim.cup_id,
+        "table": sim.table_id,
+        "tray": sim.tray_id,
+        "shelf": sim.shelf_id,
+    }[object_name]
+
+
+def get_surface_ids(sim: PyBulletEnv) -> set[int]:
+    """Get all possible surfaces in the simulator."""
+    surface_names = ["table", "tray", "shelf"]
+    return {get_object_id_from_name(n, sim) for n in surface_names}
+
+
+def get_surface_that_object_is_on(
+    object_id: int, sim: PyBulletEnv, distance_threshold: float = 1e-3
+) -> int:
+    """Get the PyBullet ID of the surface that the object is on."""
+    surfaces = get_surface_ids(sim)
+    assert object_id not in surfaces
+    object_pose = get_pose(object_id, sim.physics_client_id)
+    for surface_id in surfaces:
+        surface_pose = get_pose(surface_id, sim.physics_client_id)
+        # Check if object pose is above surface pose.
+        if object_pose.position[2] < surface_pose.position[2]:
+            continue
+        # Check for contact.
+        if check_body_collisions(
+            object_id,
+            surface_id,
+            sim.physics_client_id,
+            distance_threshold=distance_threshold,
+        ):
+            return surface_id
+    raise ValueError(f"Object {object_id} not on any surface.")
+
+
+def get_collision_ids(sim: PyBulletEnv) -> set[int]:
+    """Get all collision IDs for a sim env."""
+    return set(sim.book_ids) | {
+        sim.table_id,
+        sim.human.body,
+        sim.wheelchair.body,
+        sim.shelf_id,
+        sim.tray_id,
+        sim.side_table_id,
+    }
+
+
+def generate_side_grasps(rng: np.random.Generator) -> Iterator[Pose]:
+    """Generate side grasps."""
+    while True:
+        angle_offset = rng.uniform(-np.pi, np.pi)
+        relative_pose = get_poses_facing_line(
+            axis=(0.0, 0.0, 1.0),
+            point_on_line=(0.0, 0.0, 0),
+            radius=1e-3,
+            num_points=1,
+            angle_offset=angle_offset,
+        )[0]
+        yield relative_pose
+
+
+def generate_surface_placements(
+    surface_id: int, obj_id: int, sim: PyBulletEnv, rng: np.random.Generator
+) -> Iterator[Pose]:
+    """Sample placements uniformly on the top of the given surface."""
+    surface_extents = get_aabb_dimensions(surface_id, sim)
+    object_extents = get_aabb_dimensions(obj_id, sim)
+    placement_lb = (
+        -surface_extents[0] / 2 + object_extents[0] / 2,
+        -surface_extents[1] / 2 + object_extents[1] / 2,
+        surface_extents[2] / 2 + object_extents[2] / 2,
```

Comment:
Same as above, would this cause issues if the object is placed "inside" the surface object (like a shelf)?

Reply:
good point -- added note